File: eppt-vista/lib/Lib/hecutils.py
import math
import string
from hec.gfx2d import G2dDialog, Symbol
from hec.gfx2d import G2dPanelProp
from hec.heclib.dss import *
from hec.hecmath import TimeSeriesMath, HecMath
from hec.script import *
from java.awt import Color
from java.util import Vector
from javax.swing import *


def open_dss(dssfile, existing=True):
    """
    opens a dss files only if it exists, else fails
    """
    dss = HecDss.open(dssfile, existing)
    return dss


def close_dss(dss):
    """
    Takes a handle and closes it
    """
    dss.close()


def set_working_timewindow(dss, stime, etime):
    """
    Takes a dss handle (previously opened by call to open_dss and 
    a time window defined by strings in ddMMMyyyy HHmm format (start time and end time)  
    """
    dss.setTimeWindow(stime, etime)


def create_path_pattern(pathstr):
    """
    Creates a pattern string for the get_matching function of the form [part letter A|B|C|D|E|F]=<string to match> [space]
    from a pathstr of the format [/[part string to match| empty] repeated for A-F
    E.g. /OBS/MTZ/STAGE/1HOUR//SENSOR-36/ would be transformed to A=OBS B=MTZ C=STAGE D=1HOUR F=SENSOR-36. Note that E part is skipped as it is empty
    """
    fields = string.split(pathstr, "/")
    parts = ['A', 'B', 'C', 'D', 'E', 'F']
    pattern = ""
    for part in range(len(parts)):
        pattern = pattern + "%s=%s" % (parts[part], fields[part + 1])
    return pattern


def get_matching(dss, pattern):
    """
    Takes a dss handle and
    a pattern in format of "([part letter (A|B|C|D|E|F)]=<string to match> [space])*"
    e.g.
    get_matching(obs,'A=OBS C=MTZ E=15MIN')
    
    Fails by return None and printing message of failure.
    """
    matches = dss.getCatalogedPathnames(pattern)
    if (len(matches) >= 1):
        return dss.get(matches[0])
    else:
        print 'No match for: %s, %s' % (pattern, matches)
        return None


def plot(data, title=''):
    """
    Takes an array of data and a title and displays a plot (HECDssVue style)
    """
    plotd = newPlot(title)
    try:
        for d in data:
            plotd.addData(d)
    except:
        plotd.addData(data)
    plotd.showPlot()


def newPlot(title):
    """
    Creates a blank plot window with title
    """
    plotProp = G2dPanelProp()
    plotProp.hasToolbar = False
    return G2dDialog(None, title, False, Vector(), plotProp)


def save_as_png(p, filename):
    """
    saves plot window (G2Dialog) created by newPlot(title) method as a png graphics file
    """
    from java.awt.image import BufferedImage
    from javax.imageio import ImageIO
    from java.io import File
    bi = BufferedImage(p.size.width, p.size.height, BufferedImage.TYPE_INT_ARGB)
    g = bi.createGraphics()
    p.invalidate()
    p.validate()
    p.paint(g)
    g.dispose();
    ImageIO.write(bi, "png", File(filename));


def removeToolbar(panel):
    components = panel.getComponents();
    panel.remove(components[-1])


def create_regression_line(drun, dobs, legend):
    drunm = TimeSeriesMath(drun)
    dobsm = TimeSeriesMath(dobs)
    paired = dobsm.generateDataPairs(drunm, False)
    pairedData = paired.data
    pairedData.fullName = legend
    reg = dobsm.multipleRegression([drunm], HecMath.UNDEFINED, HecMath.UNDEFINED)
    regData = reg.data
    a = regData.yOrdinates[0][1]
    b = regData.yOrdinates[0][0]
    regData.fullName = "//REGRESSION LINE////GENERATED/"
    maxVal = drunm.max()
    minVal = drunm.min()
    regData.xOrdinates[0] = a * minVal + b
    regData.xOrdinates[1] = a * maxVal + b
    regData.yOrdinates[0][0] = minVal
    regData.yOrdinates[0][1] = maxVal
    regData.yunits = pairedData.yunits
    regData.xunits = pairedData.xunits
    regData.xtype = pairedData.xtype
    regData.ytype = pairedData.ytype
    regData.xparameter = pairedData.xparameter
    regData.yparameter = pairedData.yparameter
    regData.location = pairedData.location
    regData.version = 'LINEAR REGRESSION'
    return regData, pairedData


def scatter_plot(dobsm, drun1m, drun2m, title, legend1, legend2):
    if dobsm == None:
        return None
    from hec.lang import DSSPathString
    regData1, pairedData1 = create_regression_line(drun1m, dobsm, legend1 + " vs " + legend2)
    if drun2m != None:
        path = DSSPathString(drun2m.path)
        path.setFPart(path.getFPart() + ":2:" + obspath.getFPart())
        regData2, pairedData2 = create_regression_line(drun2m, dobsm, legend1 + " vs " + legend3)
    plots = newPlot(title);
    plots.addData(regData1)
    plots.addData(pairedData1)
    if drun2m != None:
        plots.addData(regData2)
        plots.addData(pairedData2)
    plots.showPlot()
    # plots.getViewport(regData1).getAxis("x1").setViewLimits(0,10000)
    # plots.getViewport(regData1).getAxis("y1").setViewLimits(0,10000)
    pline = plots.getCurve(regData1)
    pline.setLineColor("red")
    pline = plots.getCurve(pairedData1)
    pline.setLineVisible(0)
    pline.setSymbolType(Symbol.SYMBOL_CIRCLE)
    pline.setSymbolsVisible(1)
    pline.setSymbolSize(3)
    pline.setSymbolFillColor(pline.getLineColorString())
    pline.setSymbolLineColor(pline.getLineColorString())
    if drun2m != None:
        pline = plots.getCurve(pairedData2)
        pline.setLineColor("blue")
        pline.setLineVisible(0)
        pline.setSymbolType(Symbol.SYMBOL_SQUARE)
        pline.setSymbolsVisible(1)
        pline.setSymbolSize(3)
        pline.setSymbolFillColor(pline.getLineColorString())
        pline.setSymbolLineColor(pline.getLineColorString())
    g2dPanel = plots.getPlotpanel()
    g2dPanel.revalidate()
    g2dPanel.paintGfx()
    plots.setVisible(False)
    return g2dPanel


def save_plots_to_images(mainPanel, imageFileName, imageDir='z:/temp'):
    """
    Saves the panel used to draw plots to images
    """
    save_as_png(mainPanel, imageDir + imageFileName + ".png")


def ts_add(data):
    if data == None:
        return None
    if len(data) == 1:
        return data[0]
    else:
        result = TimeSeriesMath()
    TimeSeriesMath()


def ts_period_operation(data, interval="1DAY", operation_type="AVE"):
    """
    transforms the time series using a period operation with
    given interval (1DAY (default), 1HOUR, etc) and
    given operation type (AVE (default), MAX, MIN)
    """
    tdata = TimeSeriesMath(data).transformTimeSeries(interval, None, operation_type, 0)
    tdata.data.fullName = tdata.data.fullName + operation_type
    return tdata.data


def ts_normalize(data):
    """
    Normalize time series by dividing by data's mean
    """
    datan = data.divide(TimeSeriesMath(data).mean())
    datan.fullName = datan.fullName + "-NORMED"
    return datan.data


def calculate_weighted_rms(weights=None):
    """
    calculate the weighted rms (if weights not given assumed to be equal to 1)
    """
    rms1 = 0
    rms1_min, rms1_max = 0, 0
    rms2 = 0
    rms2_min, rms2_max = 0, 0
    rmsmap = {}
    # run2=None
    sumwts = 0
    if weights != None:
        sumwts = sumwts + weights[l]
        lrms1 = calculate_rms(drun1m.data, dobsm.data) * weights[l]
        lrms1_min = calculate_rms(drun1m_min.data, dobsm_min.data) * weights[l]
        lrms1_max = calculate_rms(drun1m_max.data, dobsm_max.data) * weights[l]
        rms1 = rms1 + lrms1
        rms1_min = rms1_min + lrms1_min
        rms1_max = rms1_max + lrms1_max
        lrms2 = calculate_rms(drun2m.data, dobsm.data) * weights[l]
        lrms2_min = calculate_rms(drun2m_min.data, dobsm_min.data) * weights[l]
        lrms2_max = calculate_rms(drun2m_max.data, dobsm_max.data) * weights[l]
        rmsmap[l] = lrms1, lrms2, lrms1_min, lrms2_min, lrms1_max, lrms2_max
        rms2 = rms2 + lrms2
        rms2_min = rms2_min + lrms2_min
        rms2_max = rms2_max + lrms2_max
    if weights != None:
        rms1 = (rms1 + rms1_min + rms1_max) / sumwts
        rms2 = (rms2 + rms2_min + rms2_max) / sumwts
        print 'RMS Run 1: %f' % rms1
        print 'RMS Run 2: %f' % rms2
        for loc in rmsmap.keys():
            print loc, rmsmap[loc]


def calculate_rms(run, obs):
    runt = TimeSeriesMath(run)
    obst = TimeSeriesMath(obs)
    tavg = obst.abs().sum() / obst.numberValidValues()
    diff = runt.subtract(obst)
    return math.fabs(math.sqrt(diff.multiply(diff).sum() / diff.numberValidValues()) / tavg) * math.log(tavg)
    # return Stats.sdev(run-obs)/Stats.avg(obs)


def arrange_plots(regplot, plotm, plotd):
    """
    arranges the plots in a certain order, 
    regplot and plotm in first row next to each other
    plotd in second row 
    """
    from javax.swing import JPanel
    from java.awt import GridBagLayout, GridBagConstraints
    mainPanel = JPanel()
    mainPanel.setLayout(GridBagLayout())
    c = GridBagConstraints()
    c.fill = c.BOTH
    c.weightx, c.weighty = 0.5, 1
    c.gridx, c.gridy, c.gridwidth, c.gridheight = 0, 0, 10, 4
    c.weightx, c.weighty = 0.5, 1
    c.gridx, c.gridy, c.gridwidth, c.gridheight = 0, 0, 10, 4
    mainPanel.add(regplot)
    c.gridx, c.gridy, c.gridwidth, c.gridheight = 0, 0, 10, 4
    c.weightx, c.weighty = 1, 1
    mainPanel.add(plotm, c)
    c.gridx, c.gridy, c.gridwidth, c.gridheight = 0, 4, 10, 6
    mainPanel.add(plotd, c)
    return mainPanel


def show_plot_in_frame(mainPanel):
    """
    embeds panel containing plots in a closable frame
    clears the panel background to white 
    """
    fr = JFrame()
    fr.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE)
    fr.getContentPane().add(mainPanel)
    fr.setSize(1100, 850);
    mainPanel.setSize(1100, 850);
    mainPanel.setBackground(Color.WHITE);
    fr.show();
    return fr


def ts_plot(data1, data2=None, data3=None, title="", legend1="", legend2="", legend3=""):
    plotd = newPlot(title)
    if data1 != None:
        plotd.addData(data1)
    if data2 != None:
        plotd.addData(data2)
    if data3 != None:
        plotd.addData(data3)
    plotd.showPlot()
    plotd.setPlotTitleText(title)
    plotd.setPlotTitleVisible(True)
    legend_label = plotd.getLegendLabel(data1)
    legend_label.setText(legend1)
    if data2 != None:
        legend_label = plotd.getLegendLabel(data2)
        legend_label.setText(legend2)
    if data3 != None:
        legend_label = plotd.getLegendLabel(data3)
        legend_label.setText(legend3)
    plotd.setVisible(False)
    xaxis = plotd.getViewport(0).getAxis("x1")
    # vmin =xaxis.getViewMin()+261500. # hardwired to around july 1, 2008 ???
    # xaxis.setViewLimits(vmin,vmin+10000.)
    if data1 != None:
        pline = plotd.getCurve(data1)
        pline.setLineVisible(1)
        pline.setLineColor("blue")
        pline.setSymbolType(Symbol.SYMBOL_CIRCLE)
        pline.setSymbolsVisible(0)
        pline.setSymbolSize(3)
        pline.setSymbolSkipCount(0)
        pline.setSymbolFillColor(pline.getLineColorString())
        pline.setSymbolLineColor(pline.getLineColorString())
    if data3 != None:
        pline = plotd.getCurve(data3)
        pline.setLineVisible(1)
        pline.setLineColor("orange")
    g2dPanel = plotd.getPlotpanel()
    g2dPanel.revalidate();
    g2dPanel.paintGfx();
    dpanel = plotd.getPlotpanel()
    return dpanel
    # removeToolbar(dpanel)

File: eppt-vista/src/main/java/gov/ca/hec/HecUtils.java
/*
 * Copyright (c) 2019
 * California Department of Water Resources
 * All Rights Reserved.  DWR PROPRIETARY/CONFIDENTIAL.
 * Source may not be released without written approval from DWR
 */

package gov.ca.hec;

import java.util.Vector;
import java.util.regex.Pattern;

import hec.heclib.dss.CondensedReference;
import hec.heclib.dss.DSSPathname;
import hec.heclib.dss.HecDss;
import hec.heclib.util.HecTime;
import hec.hecmath.HecMath;
import hec.hecmath.HecMathException;
import hec.hecmath.TimeSeriesMath;
import hec.io.DataContainer;
import hec.io.TimeSeriesContainer;

/**
 * A few utility functions to access HECDSSVue functionality
 * 
 * @author psandhu
 *
 */
public class HecUtils {
	/**
	 * opens a dss files only if it exists, else fails
	 * 
	 * @throws Exception
	 */
	public static HecDss openDSS(String dssfile) throws Exception {
		return HecDss.open(dssfile, true);
	}

	/**
	 * opens a dss files only if it exists, else fails
	 * 
	 * @throws Exception
	 */
	public static HecDss openDSS(String dssfile, boolean existing) throws Exception {
		return HecDss.open(dssfile, existing);
	}

	/**
	 * Takes a handle and closes it
	 * 
	 * @param dss
	 */
	public static void closeDSS(HecDss dss) {
		dss.close();
	}

	/**
	 * Takes a dss handle (previously opened by call to open_dss and a time
	 * window defined by strings in ddMMMyyyy HHmm format (start time and end
	 * time) Note: setting these windows before accessing dss file can lead to
	 * some odd truncation of data.
	 */
	public static void setWorkingTimewindow(HecDss dss, String stime, String etime) {
		dss.setTimeWindow(stime, etime);
	}

	/**
	 * Creates a pattern string for the get_matching function of the form [part
	 * letter A|B|C|D|E|F]=<string to match> [space] from a pathstr of the
	 * format [/[part string to match| empty] repeated for A-F E.g.
	 * /OBS/MTZ/STAGE/1HOUR//SENSOR-36/ would be transformed to A=OBS B=MTZ
	 * C=STAGE D=1HOUR F=SENSOR-36. Note that E part is skipped as it is empty
	 * 
	 * @return
	 */
	public static String createPathPattern(String pathstr) {
		String[] fields = pathstr.split("/");
		String[] parts = { "A", "B", "C", "D", "E", "F" };
		String pattern = "";
		for (int part = 0; part < parts.length; part++) {
			String partVal = fields[part + 1].trim();
			if (partVal.length() == 0)
				continue;
			pattern = pattern + parts[part] + "=" + partVal + " ";
		}
		return pattern.trim();
	}

	/**
	 * Takes a dss handle and a pattern in format of "([part letter
	 * (A|B|C|D|E|F)]=<string to match> [space])*" e.g. get_matching(obs,'A=OBS
	 * C=MTZ E=15MIN')
	 * 
	 * Fails by return None and printing message of failure.
	 * 
	 * @throws Exception
	 */
	public static DataContainer getMatching(HecDss dss, String pattern) throws Exception {
		@SuppressWarnings("unchecked")
		Vector<String> matches = dss.getCatalogedPathnames(pattern);
		if (matches.size() >= 1) {
			return dss.get(matches.get(0));
		} else {
			System.err.println("No match for: " + pattern + ", " + matches);
			return null;
		}
	}

	/**
	 * Calculate root mean square error between the two time series and return the value.
	 * @throws HecMathException
	 * 
	 */
	public static double calculateRMS(TimeSeriesContainer model, TimeSeriesContainer obs) throws HecMathException {
		TimeSeriesMath modelt = new TimeSeriesMath(model);
		TimeSeriesMath obst = new TimeSeriesMath(obs);
		HecMath diff = modelt.subtract(obst);
		return Math.sqrt(diff.multiply(diff).sum() / diff.numberValidValues());
	}

	/**
	 * Calculate Nash Sutcliffe efficiency between the two time series.
	 * <a href="https://en.wikipedia.org/wiki/Nash%E2%80%93Sutcliffe_model_efficiency_coefficient">Nash Sutcliffe Efficiency Coefficient</a>
	 * 
	 */
	public static double calculateNashSutcliffeEfficiency(TimeSeriesContainer model, TimeSeriesContainer obs)
			throws HecMathException {
		TimeSeriesMath modelt = new TimeSeriesMath(model);
		TimeSeriesMath obst = new TimeSeriesMath(obs);
		double tavg = obst.abs().sum() / obst.numberValidValues();
		HecMath diffModel = modelt.subtract(obst);
		HecMath diffObs = obst.subtract(tavg);
		return 1 - (diffModel.multiply(diffModel)).sum() / (diffObs.multiply(diffObs)).sum();
	}

	/**
	 *  For the open dss handle and dss path and time window string return the time series container for the request
	 */
	public static DataContainer getMatching(HecDss dss, String dsspath, String twStr) throws Exception {
		String dsspathRegEx = searchRegex(replaceAndEscape(dsspath), dss);
		if (dsspathRegEx == null) {
			return null;
		}
		String startTime = null;
		String endTime = null;
		if (twStr == null || twStr.length() == 0) {
			HecTime start = new HecTime();
			HecTime end = new HecTime();
			dss.getTimeSeriesExtents(dsspathRegEx, start, end);
			startTime = start.dateAndTime();
			endTime = end.dateAndTime();
		} else {
			String[] twFields = twStr.split(" ");
			startTime = twFields[0] + " " + twFields[1];
			endTime = twFields[2] + " " + twFields[3];
		}
		DataContainer dataContainer = dss.get(dsspathRegEx, startTime, endTime);
		return dataContainer;
	}

	/**
	 * parse time window from dss path. This is the format that the condensed catalog from HECDSS files of the form 
	 * 01JAN1990 2400-05JAN1992 0200 
	 */
	public static String parseTimeWindowFromPath(String dsspath) {
		DSSPathname pathname = new DSSPathname(dsspath);
		String[] fields = pathname.getDPart().split(" - ");
		return fields[0] + " 0000 " + fields[1] + " 0000";
	}

	/**
	 * Search for regular expression in the dss condensed catalog. 
	 * @ return first pathname that matches or null
	 */
	public static String searchRegex(String regpath, HecDss dss) {
		Vector<CondensedReference> condensedCatalog = dss.getCondensedCatalog();
		Pattern p = Pattern.compile(regpath, Pattern.CASE_INSENSITIVE);
		for (CondensedReference condensedReference : condensedCatalog) {
			String pathname = condensedReference.getNominalPathname();
			if (p.matcher(pathname).matches()) {
				return pathname;
			}
		}
		return null;
	}
	
	/**
	 * Escaping certain characters in a regular expression for the dss pathname and returning as a string
	 */
	public static String replaceAndEscape(String dsspath) {
		if (dsspath == null)
			return ".*";
		String[] fields = dsspath.split("/");
		StringBuilder sb = new StringBuilder();
		int slashCount = 0;
		for (String f : fields) {
			slashCount++;
			if (slashCount == 1) {
				sb.append("/");
				continue;
			}
			if (f.length() == 0) {
				sb.append(".*");
			} else {
				sb.append(Pattern.quote(f));
			}
			sb.append("/");
		}
		while (slashCount < 7) {
			sb.append(".*/");
			slashCount++;
		}
		return sb.toString();
	}

	/**
	 * Average the time series for the given interval (as a string) and return time series
	 * @param dataContainer
	 * @param intervalAsString
	 *            e.g., "1DAY" for daily averaging
	 * @return
	 * @throws HecMathException
	 */
	public static DataContainer average(DataContainer dataContainer, String intervalAsString) throws HecMathException {
		return new TimeSeriesMath((TimeSeriesContainer) dataContainer)
				.transformTimeSeries(intervalAsString, null, "AVE").getData();
	}

}
